### PR TITLE
FEATURE: New excluded classes setting.

### PR DESCRIPTION
**Pull request by danielhollas**
As per user request on meta: https://meta.discourse.org/t/linkify-words-in-post/82193/160?u=danekhollas

By default, we will not autolink inside oneboxes.

The easiest way to test this is in quotes, by adding `quote` as an excluded class.

image

FYI @SamSaffron

### CODE REVIEW

File: javascripts/discourse-linkify/lib/utilities.js.es6 [hunk 1]
```diff
@@ -126,19 +126,33 @@ const modifyText = function(text, action) {
...
+
+  if (!classes) return false;
+
+  // Return true if at least one of the classes should be skipped
+  return classes.split(" ").some(cls => {
+    for (let skipClass in skipClasses) {
+      if (cls === skipClass) return true;
+    }
+    return false;
+  });
```
Comment (ZogStriP) [inline javascripts/discourse-linkify/lib/utilities.js.es6]: Isn't `skipClasses` an hash? You should be able to do much shorter 😉 

```suggestion
  return classes.split(" ").some(cls => cls in skipClasses);
```

Comment (danielhollas) [reply javascripts/discourse-linkify/lib/utilities.js.es6]: Ah, of course. :man_facepalming: Thanks! This also exposed a silly bug that I just fixed.

File: javascripts/discourse-linkify/initializers/initialize-discourse-linkify.js.es6 [hunk 2]
```diff
@@ -40,12 +49,12 @@ export default {
...
-        actions.forEach(a => {
-          if (Object.keys(a.inputs).length > 0) {
-            modifyNode($elem[0], a, skipTags)
+        actions.forEach(action => {
+          if (Object.keys(action.inputs).length > 0) {
+            traverseNodes($elem[0], action, skipTags, skipClasses)
           }
         });
-      });
+      }, {'id': 'linkify-words-theme'});
```
Comment (danielhollas) [inline javascripts/discourse-linkify/initializers/initialize-discourse-linkify.js.es6]: I added an 'id' to the api.decorateCooked call as per

javascripts/discourse-linkify/initializers/initialize-discourse-linkify.js.es6

Is that okay?

File: javascripts/discourse-linkify/lib/utilities.js.es6 [hunk 3]
```diff
@@ -126,19 +126,26 @@ const modifyText = function(text, action) {
   }
 }
 
-const modifyNode = function(elem, action, skipTags) {
+const isSkippedClass = function(classes, skipClasses) {
+  if (!classes) return false;
+  // Return true if at least one of the classes should be skipped
+  return classes.split(" ").some(cls => cls in skipClasses);
```
Comment (ZogStriP) [inline javascripts/discourse-linkify/lib/utilities.js.es6]: You can do this in one-line 😉 

```suggestion
  // Return true if at least one of the classes should be skipped
  return classes && classes.split(" ").some(cls => cls in skipClasses);
```

File: javascripts/discourse-linkify/initializers/initialize-discourse-linkify.js.es6 (hunk 2 above)
Comment (ZogStriP) [reply javascripts/discourse-linkify/initializers/initialize-discourse-linkify.js.es6]: Yup, and `id` is always a good idea 👍

**Comment by ZogStriP** [comment]
It's looking good 👍 

Any chances you could add some tests to at least ensure basic functionality is working?

**Comment by danielhollas** [comment]
I agree this theme component badly needs tests (there are tons of corner cases that tricked me in previous PRs) but currently there are none and I have never written tests for Discourse themes/plugins. Feels like good opportunity for learning, but I'd propose to leave it to a future PR? (will get to it next weekend earliest). Also, I haven't seen any theme components with tests yet, would be glad for links...

**Comment by ZogStriP** [comment]
Thanks for your work @danielhollas 👏

**Comment by danielhollas** [comment]
Thanks for the review, @ZogStriP :heart:
